lib/publish.js

```diff
@@ -1,167 +1,105 @@
-  }).then((pkg) => {
-    return lifecycle(pkg, 'prepublishOnly', arg)
-  }).then(() => {
-    return readJson(path.join(arg, 'package.json'))
-  }).then((pkg) => {
-    return cacache.tmp.withTmp(opts.tmp, {tmpPrefix: 'fromDir'}, (tmpDir) => {
-      const target = path.join(tmpDir, 'package.tgz')
-      return pack.packDirectory(pkg, arg, target, null, true)
-        .tap((c) => { contents = c })
-        .then((c) => !opts.json && pack.logContents(c))
-        .then(() => upload(pkg, false, target, opts))
-    })
-  }).then(() => {
-    return readJson(path.join(arg, 'package.json'))
-  }).tap((pkg) => {
-    return lifecycle(pkg, 'publish', arg)
-  }).tap((pkg) => {
-    return lifecycle(pkg, 'postpublish', arg)
-  })
-    .then(() => contents)
-}
+  let manifest = await readJson(`${arg}/package.json`)
 
-function publishFromPackage (arg, opts) {
-  return cacache.tmp.withTmp(opts.tmp, {tmpPrefix: 'fromPackage'}, tmp => {
-    const extracted = path.join(tmp, 'package')
-    const target = path.join(tmp, 'package.json')
-    return tarball.toFile(arg, target, opts)
-      .then(() => extract(arg, extracted, opts))
-      .then(() => readJson(path.join(extracted, 'package.json')))
-      .then((pkg) => {
-        return BB.resolve(pack.getContents(pkg, target))
-          .tap((c) => !opts.json && pack.logContents(c))
-          .tap(() => upload(pkg, false, target, opts))
-      })
+  // prepublishOnly
+  await runScript({
+    event: 'prepublishOnly',
+    path: arg,
+    stdio: 'inherit',
+    pkg: manifest
   })
-}
+  
+  const tarballData = await pack(arg)
+  const pkgContents = await getContents(manifest, tarballData)
 
-function upload (pkg, isRetry, cached, opts) {
-  if (!opts.dryRun) {
-    return readFileAsync(cached).then(tarball => {
-      return otplease(opts, opts => {
-        return libpub(pkg, tarball, opts)
-      }).catch(err => {
-        if (
-          err.code === 'EPUBLISHCONFLICT' &&
-          opts.force &&
-          !isRetry
-        ) {
-          log.warn('publish', 'Forced publish over ' + pkg._id)
-          return otplease(opts, opts => libunpub(
-            npa.resolve(pkg.name, pkg.version), opts
-          )).finally(() => {
-            // ignore errors.  Use the force.  Reach out with your feelings.
-            return otplease(opts, opts => {
-              return upload(pkg, true, tarball, opts)
-            }).catch(() => {
-              // but if it fails again, then report the first error.
-              throw err
-            })
-          })
-        } else {
-          throw err
-        }
-      })
+  if (!opts.json) {
+    logTar(pkgContents, {
+      log,
+      unicode: npm.config.get('unicode')
     })
-  } else {
-    return opts.Promise.resolve(true)
   }
+  
+  try {
+    if (!opts.dryRun) {
+      await otplease(opts, opts => libpub(arg, manifest, opts))
+    }
+  } catch (err) {
+    throw err
+  }
+
+  manifest = await readJson(`${arg}/package.json`)
```

Comment:
I _don't_ think we have to re-read it here.  The purpose of re-reading the manifest is in case it changed, so that we send the latest and greatest thing to the registry.

(Also, maybe this is the _only_ place we need to re-read it?  Since we don't put the whole package contents into the environment any more, anyone looking to mess with it in a script will be loading it themselves anyway.)

Reply:
So, since we don't have to re-load it unless we're actually publishing, the previous try/catch and this line resolve to:

```js
if (!opts.dryRun) {
  manifest = await readJson(`${arg}/package.json`)
  await otplease(opts, opts => libpub(arg, manifest, opts))
}
```